fl4health/utils/dataset_converter.py

```diff
@@ -0,0 +1,159 @@
+from abc import ABC
+from typing import Callable, Optional, Tuple, Union
+
+import torch
+
+from fl4health.utils.dataset import BaseDataset
+
+
+class DatasetConverter(ABC, BaseDataset):
+    """Dataset converter classes are designed to re-format any dataset for a given training task,
+    and to fit it into the unified training scheme of supervised learning in clients.
+    Converters can be used in the data loading step. They can also apply a light
+    pre-processing step on datasets before the training process.
+    """
+
+    def __init__(self, converter_function: Callable, dataset: Union[None, BaseDataset]) -> None:
+        self.converter_function = converter_function
+        self.dataset = dataset
+
+    def __getitem__(self, index: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Overriding this function from BaseDataset allows the converter to be compatible with the data transformers.
+        # converter_function is applied after the transformers.
+        assert self.dataset is not None, "Error: no dataset is set, use convert_dataset(your_dataset: BaseDataset)"
+        data, target = self.dataset.__getitem__(index)
+        return self.converter_function(data, target)
+
+    def __len__(self) -> int:
+        assert self.dataset is not None, "Error: dataset is should be either converted or initiated."
+        return len(self.dataset)
+
+    def convert_dataset(self, dataset: BaseDataset) -> BaseDataset:
+        """Applies the converter function over the dataset when the dataset is used
+        (i.e. during the dataloader creation step)."""
+        # Dataset can be added/changed at any point in the pipeline.
+        self.dataset = dataset
+        # Returning this object as the converted dataset since this class overrides
+        # the __getitem__ function of BaseDataset.
+        return self
+
+
+class AutoEncoderDatasetConverter(DatasetConverter):
+    """A dataset converter specific to formating supervised data such as MNIST for
+    self-supervised training in autoencoder-based models, and potentially handling
+    the existence of additional input i.e. condition.
+    This class includes three converter functions that are chosen based on the condition,
+    other converter functions can be added or passed to support other conditions.
+    """
+
+    def __init__(
+        self,
+        condition: Union[None, str, torch.Tensor],
+        custom_converter_function: Optional[Callable] = None,
```

Comment:
Do we have an example using this? The condition tensor here is for a fixed conditional vector. Something akin to a client id, as you have been using. I imagine the custom_converter_function could be used to consider the encoder based on dynamic vectors (similar to label but derived from the data). For example, age group + gender (young, adult, old, male, female) might be encoded for every example such that the encoder/decoder knows this when embedding or synthetically generating.

Reply:
Yes, that is exactly the goal. The user can define a fixed condition tensor for each client, or define how the condition for each data sample should be created (your example) via passing a custom_converter_function. If the user defines a custom converter function, that converter will be used to define the dynamic condition based on each data sample. This is somehow similar to the `_cat_input_label()` which is used if the condition is 'label'. I have tested it in our tests.
So I would say this feature is supported hopefully, but the user needs to define the custom converter function in their client code. We can definitely add more examples if needed.

Reply:
That makes sense! Perhaps we can just convert either the convolutional or mlp based CVAE training example we have to do this? I know there isn't "side information" to the MNIST dataset, but maybe we could artificially create some, where we encode the labels in a fixed/random dense vector of dimension three or something as a mock example? Just a thought.

Reply:
In `conv_cvae_example`, now we have a custom converter function (`binary_class_condition_data_converter`) designed to create a dense vector of the binary representation of each sample's class. This can be considered as a custom non-fixed conditioning.